### PR TITLE
check file content md5

### PR DESCRIPTION
**Pull request by lfbdev**
# Summary

add md5 check on local file and remote file

### Related Issues

#103  add content-md5 check on local file and remote file

### CODE REVIEW

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 1]
```diff
@@ -114,3 +114,24 @@ func createPathIfNotExists(path string) error {
 	}
 	return nil
 }
+
+// fetchContentMD5 get Content-MD5
+func (pd *PbDownloadClient) fetchContentMD5(pluginFilename, version string) (string, error) {
+	downloadURL := fmt.Sprintf("%s/v%s/%s", defaultReleaseUrl, version, pluginFilename)
+	pd.Timeout = time.Second * 60 * 60
+	resp, err := pd.Get(downloadURL)
```
Comment (IronCore864) [inline internal/pkg/pluginmanager/pbdownloader.go:122]: Check if we need to close the response.

File: internal/pkg/pluginmanager/manager.go [hunk 2]
```diff
@@ -33,8 +34,34 @@ func DownloadPlugins(conf *configloader.Config) error {
 			}
 			continue
 		}
+		// check md5
+		if dup, _ := checkFileMD5(filepath.Join(pluginDir, pluginFileName), dc, pluginFileName, tool.Version); !dup {
```
Comment (IronCore864) [inline internal/pkg/pluginmanager/manager.go:38]: Handle error.

Comment (lfbdev) [reply internal/pkg/pluginmanager/manager.go:38]: err check added

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 1 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go:123]: add defer to close response conn

**Review by IronCore864 — APPROVED** [review]

File: internal/pkg/pluginmanager/manager.go [hunk 3]
```diff
@@ -33,8 +34,40 @@ func DownloadPlugins(conf *configloader.Config) error {
 			}
 			continue
 		}
+		// check md5
+		dup, err := checkFileMD5(filepath.Join(pluginDir, pluginFileName), dc, pluginFileName, tool.Version)
+		if err != nil {
+			return err
+		}
+
+		if !dup {
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/manager.go:43]: if dup {
   log
   continue
}
code


let `if` shorter

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 4]
```diff
@@ -114,3 +114,24 @@ func createPathIfNotExists(path string) error {
...
+// fetchContentMD5 get Content-MD5
+func (pd *PbDownloadClient) fetchContentMD5(pluginFilename, version string) (string, error) {
+	downloadURL := fmt.Sprintf("%s/v%s/%s", defaultReleaseUrl, version, pluginFilename)
+	pd.Timeout = time.Second * 60 * 60
+
+	resp, err := pd.Get(downloadURL)
+	if err != nil {
+		return "", err
+	}
+	if resp != nil {
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:127]: Maybe the `if` here is redundant

**Review by daniel-hutao — CHANGES_REQUESTED** [review]
All feedbacks from me.

File: internal/pkg/pluginmanager/manager.go [hunk 5]
```diff
@@ -33,8 +34,39 @@ func DownloadPlugins(conf *configloader.Config) error {
...
+			return err
+		}
+
+		if dup {
+			log.Printf("Plugin: %s already exists, no need to download.", pluginFileName)
+			continue
+		}
+
+		log.Printf("Plugin: %s changed and will be overrided.", pluginFileName)
+		os.Remove(filepath.Join(pluginDir, pluginFileName))
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/manager.go:49]: error handle

File: internal/pkg/pluginmanager/manager.go [hunk 6]
```diff
@@ -33,8 +34,39 @@ func DownloadPlugins(conf *configloader.Config) error {
...
+	localmd5, err := LocalContentMD5(file)
+	if err != nil {
+		return false, err
+	}
+	remotemd5, err := dc.fetchContentMD5(pluginFileName, version)
+	if err != nil {
+		return false, err
+	}
+
+	if strings.Compare(localmd5, remotemd5) == 0 {
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/manager.go:68]: It is usually clearer and always faster to use the built-in
// string comparison operators ==, <, >, and so on.

**Review by daniel-hutao — CHANGES_REQUESTED** [review]
2 tips added

**Review by daniel-hutao — APPROVED** [review]

**Pull request merged**